ENH: capture all errors logged by gdal when opening a file fails

theroggy opened the pull request:
This was present in Rasterio (https://github.com/rasterio/rasterio/pull/2526) and has been included in fiona::_err recently (https://github.com/Toblerity/Fiona/pull/1408)... so this PR is based on the change there.

Closes #491

jorisvandenbossche commented on pyogrio/_err.pyx:
```diff
@@ -1,11 +1,25 @@
+    CPLErr, CPLErrorHandler, CPLDefaultErrorHandler, CPLPopErrorHandler,
+    CPLPushErrorHandler, CPLPushErrorHandlerEx)
+
+log = logging.getLogger(__name__)
```
Do we need to logging part of this? (in contrast to fiona, we in general don't enable logging)

theroggy replied:
Not sure how interesting the logging is in general, but as it was apparently consciously removed when porting from fiona I suppose it's not that interesting, so I'll remove it.

jorisvandenbossche commented on pyogrio/_err.pyx:
```diff
@@ -173,8 +245,31 @@ cdef inline object exc_check():
         return
 
 
+cdef get_last_error_msg():
```
This is not actually used yet? Above in the new code you directly use `CPLGetLastErrorMsg()`, but I suppose that could use this helper?

theroggy replied:
True. I extracted the error message cleaning code in a new function `clean_error_message` and now use it where applicable.

jorisvandenbossche commented on pyogrio/_err.pyx:
```diff
@@ -248,3 +344,197 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+
+
+IF UNAME_SYSNAME == "Windows":
+    cdef void __stdcall stacking_error_handler(
```
Is the only difference this `__stdcall`? And any idea why his is needed? (we don't have it for the `error_handler` defined above)

theroggy replied:
Yes, that's the only difference. I had a quick further look, and apparently `__stdcall` is even ignored by ARM and x64 compilers, so there is probably not a lot of use in keeping it, so I removed it.

brendan-ward commented on pyogrio/_err.pyx:
```diff
@@ -248,3 +344,197 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+        """Wrap a GDAL/OGR function that returns CPLErr (int).
+
+        Raises an exception if a non-fatal error has been added to the
+        exception stack.
```
```suggestion
        exception stack.  Only checks for errors if `err` is set to a nonzero value.
```

brendan-ward commented on pyogrio/_err.pyx:
```diff
@@ -248,3 +344,197 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+        """Wrap a GDAL/OGR function that returns a pointer.
+
+        Raises an exception if a non-fatal error has been added to the
+        exception stack.
```
```suggestion
        exception stack.  Only checks for errors if `ptr` is `NULL`.
```

brendan-ward commented on pyogrio/_err.pyx:
```diff
@@ -248,3 +344,197 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+cpl_errs = GDALErrCtxManager()
+
+
+cdef class StackChecker:
```
```suggestion
cdef class ErrorHandler:
```

The stack part seems an implementation detail, I think the part we want to focus here on the name for usage below is that it checks for errors.

brendan-ward commented on pyogrio/_err.pyx:
```diff
@@ -248,3 +344,197 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+    def __init__(self, error_stack=None):
+        self.error_stack = error_stack or {}
+
+    cdef int exc_wrap_int(self, int err) except -1:
```
```suggestion
    cdef int check_int(self, int err) except -1:
```

I wonder if it is time for us to diverge from Fiona's original naming here and try to make this a bit more clear.
(also good to consider doing this for the top-level function names too.)

theroggy replied:
ok, renamed

brendan-ward commented on pyogrio/_err.pyx:
```diff
@@ -248,3 +344,197 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+
+        return err
+
+    cdef void *exc_wrap_pointer(self, void *ptr) except NULL:
```
```suggestion
    cdef void *check_pointer(self, void *ptr) except NULL:
```

theroggy replied:
Ok, renamed.

brendan-ward commented on pyogrio/_err.pyx:
```diff
@@ -248,3 +344,197 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+
+
+@contextlib.contextmanager
+def stack_errors():
```
```suggestion
def get_error_handler():
```

The purpose of the context manager is to get the error handler.

brendan-ward commented on pyogrio/_io.pyx:
```diff
@@ -196,28 +205,33 @@ cdef void* ogr_open(const char* path_c, int mode, char** options) except NULL:
-        )
-
-        return ogr_dataset
+        with stack_errors() as error_stack_checker:
```
```suggestion
        with get_error_handler() as error_handler:
```

(to demonstrate increased clarity of above suggestions)

theroggy replied:
I renamed the context handler to `capture_errors` instead of `get_error_handler` as I think this is a clearer description of the behaviour.

OK?

brendan-ward commented on pyogrio/_io.pyx:
```diff
@@ -196,28 +205,33 @@ cdef void* ogr_open(const char* path_c, int mode, char** options) except NULL:
-        return ogr_dataset
+        with stack_errors() as error_stack_checker:
+            ogr_dataset = GDALOpenEx(path_c, flags, NULL, <const char *const *>options, NULL)
+            return error_stack_checker.exc_wrap_pointer(ogr_dataset)
```
```suggestion
            return error_handler.check_pointer(ogr_dataset)
```
(ditto above)

brendan-ward reviewed (commented):
Thanks for working on this @theroggy !  A few suggestions re: naming things.

theroggy commented:
> Thanks for looking into this! Added a few quick questions
> 
> Do you have a link to the PRs that added this to rasterio/fiona (that might have some background)?

I added links to them in the [PR description](https://github.com/geopandas/pyogrio/pull/495#issue-2647800422)...

jorisvandenbossche commented on pyogrio/_err.pyx:
```diff
@@ -1,25 +1,41 @@
+"""
+
+import contextlib
+import logging
```
```suggestion
```

jorisvandenbossche commented on pyogrio/_err.pyx:
```diff
@@ -132,49 +159,114 @@ exception_map = {
+    def __exit__(self, exc_type=None, exc_val=None, exc_tb=None):
+        cdef int err_type = CPLGetLastErrorType()
+        cdef int err_no = CPLGetLastErrorNo()
+        cdef const char *msg = CPLGetLastErrorMsg()
```
Should this also use the `get_last_error_msg()` helper?

theroggy replied:
In the stack, they are still c const char* 's... it is only when the stack is being interpreted that the messages are converted to python objects...

theroggy replied:
This context handler isn't used anywhere in pyogrio, and thinking further on it I'm also not sure why there a two gdal error context handlers in fiona/rasterio so I removed it.

jorisvandenbossche commented on pyogrio/_err.pyx:
```diff
@@ -132,49 +159,114 @@ exception_map = {
+    """A manager for GDAL error handling contexts."""
+
+    def __enter__(self):
+        CPLErrorReset()
```
Should this also do a `CPLErrorReset()` at the end of `__exit__`?

theroggy replied:
Code removed...

jorisvandenbossche commented on pyogrio/_err.pyx:
```diff
@@ -132,49 +159,114 @@ exception_map = {
+        cdef const char *msg = CPLGetLastErrorMsg()
+        # TODO: warn for err_type 2?
+        if err_type >= 2:
+            raise exception_map[err_no](err_type, err_no, msg)
```
We use `exception_map.get(err_no, CPLE_BaseError)` (for the case that there is an err_no we are not aware of, maybe also for future compatibility?), should we use that here as well?

theroggy replied:
Code removed...

jorisvandenbossche commented on pyogrio/_err.pyx:
```diff
@@ -248,3 +339,121 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+
+    # stacking_error_handler records GDAL errors in the order they occur and
+    # converts them to exceptions.
+    CPLPushErrorHandlerEx(<CPLErrorHandler>stacking_error_handler, NULL)
```
If we would (eventually?) handle all catching of GDAL errors using this context manager, then we don't need to each time register this error handler and pop it again?

theroggy replied:
True. I'm not sure why both approaches (or even three taking the second context handler in account) are being used in rasterio/fiona...

In fiona at least it is only used when files are opened, all other calls use the getlasterror approach... possibly there is a downside to capturing all errors in some cases?

jorisvandenbossche replied:
> I'm not sure why both approaches are being used in rasterio/fiona...

That might also be partly just historical reasons, since this was only added more recently to fiona (and although it could be used everywhere, there was maybe also no pressing need to change the code that much?)

theroggy commented on pyogrio/_err.pyx:
```diff
@@ -1,25 +1,40 @@
 
 
 # CPL Error types as an enum.
 class GDALError(IntEnum):
+    """GDAL error types/classes.
+    
+    GDAL doc: https://gdal.org/en/latest/doxygen/cpl__error_8h.html#a463ba7c7202a505416ff95b1aeefa2de
+    """
     none = CE_None
     debug = CE_Debug
     warning = CE_Warning
     failure = CE_Failure
     fatal = CE_Fatal
```
This enum isn't use anywhere, should it be kept here or shall I remove it?

brendan-ward replied:
Removing it is fine, likely a leftover from porting from Fiona in the early days.

jorisvandenbossche commented on pyogrio/_err.pyx:
```diff
@@ -132,53 +144,88 @@ exception_map = {
-        return exception_map.get(
-            err_no, CPLE_BaseError)(err_type, err_no, msg)
+    else:
+        msg = ""
```
Do we want to keep "No error message." here? Or we don't want that for the case that we are stacking multiple errors where not every error might have a message?

theroggy replied:
That's why I retained the empty string. I'm not sure if having no error message actually happens... it seems there is no test for it either...

theroggy replied:
I moved the check so "No error message." will be shown in the typical cases where there would be an exception without message...

jorisvandenbossche commented on pyogrio/_err.pyx:
```diff
@@ -187,40 +234,45 @@ cdef void *exc_wrap_pointer(void *ptr) except NULL:
-cdef int exc_wrap_ogrerr(int err) except -1:
-    """Wrap a function that returns OGRErr (int) but does not use the
-    CPL error stack.
+cdef int check_ogrerr(int err) except -1:
```
Maybe for another PR, but I am wondering if we should keep this helper function. 
At the moment we use this for calling `OGR_L_GetExtent`, but then if the return value is non-zero, that indicates the bounds are not known and we catch the error to set the bounds to `None`. But so that checking for non-zero return we can do directly where we call `OGR_L_GetExtent`, instead of raising a python error and catching that again to do the same.

And the other place it is used is in `exc_wrap_ogrerr(exc_wrap_int(OSRImportFromEPSG(srs, 4326)))`, which seems a bit fishy in that it chains both exception handlers. I don't think the `exc_wrap_ogrerr` has any effect there.

theroggy replied:
In both cases the added value is indeed not very obvious...

theroggy replied:
I'll remove it.

jorisvandenbossche commented on pyogrio/_err.pyx:
```diff
@@ -248,3 +298,127 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+    if any were captured.
+    """
+    CPLErrorReset()
+    global _ERROR_STACK
```
What is the reason this needs to be a global variable? Because otherwise `stacking_error_handler` cannot access it? (being a `cdef` function, we would not be able to pass a python object as argument it?)

jorisvandenbossche replied:
Ah, and also of course the `stacking_error_handler` has a fixed signature and gets called from GDAL, so if we want to pass some custom data to that function, that's what the user data could be used for in `CPLPushErrorHandlerEx` (and retrieved with `CPLGetErrorHandlerUserData`).

I suppose we could use that to pass through the error stack, but at that point using a global variable might be simpler

jorisvandenbossche replied:
But, isn't the point of a context var that you don't need to use `global` for it to work? Does it still work if you just remove this line?

theroggy replied:
I had a quick read, and indeed... I'd guess as well that the `global ...` statements shouldn't be there. I remover them, and the tests still pass without issues.

jorisvandenbossche commented on pyogrio/_err.pyx:
```diff
@@ -248,3 +298,127 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+
+    # stacking_error_handler records GDAL errors in the order they occur and
+    # converts them to exceptions.
+    CPLPushErrorHandlerEx(<CPLErrorHandler>stacking_error_handler, NULL)
```
```suggestion
    CPLPushErrorHandler(<CPLErrorHandler>stacking_error_handler)
```

jorisvandenbossche commented on pyogrio/_io.pyx:
```diff
@@ -196,28 +205,33 @@ cdef void* ogr_open(const char* path_c, int mode, char** options) except NULL:
 
     except CPLE_BaseError as exc:
-        if str(exc).endswith("a supported file format."):
+        # If there are inner exceptions, append their errmsg to the errmsg
+        errmsg = str(exc)
+        inner = exc.__cause__
+        while inner is not None:
+            errmsg = f"{errmsg}; {inner}"
+            inner = inner.__cause__
```
Is this not something we want to do in the logic in `_err.pyx` ?

theroggy replied:
I've been wondering about this as well.

I'll add a parameter "squash_errors" or something like to the `check_...` functions that to have the choice in having the errors returned as a stack of exceptions or as one exception with all error messages "squashed" to one.

theroggy replied:
Done.

jorisvandenbossche reviewed (commented):
Thanks! A few more comments/questions

jorisvandenbossche commented on pyogrio/_err.pyx:
```diff
@@ -1,25 +1,26 @@
+    CPLErr, CPLErrorHandler, CPLDefaultErrorHandler, CPLPopErrorHandler,
+    CPLPushErrorHandler, CPLPushErrorHandlerEx)
 
+_ERROR_STACK = ContextVar("error_stack")
```
Do you know the reason for using a ContextVar here? (and not a thread-local variable) from quickly reading up on it, it seems this is mostly used to have something similar to thread-local storage but in async / coroutines?

theroggy replied:
No, I didn't really look into it in that detail...

brendan-ward commented:
This appears to be raising a lot of test errors on GDAL 3.4.3 specifically; is that expected?

theroggy commented:
> This appears to be raising a lot of test errors on GDAL 3.4.3 specifically; is that expected?

No... I've had a quick look at it... but I didn't immediately see why these errors occur...

theroggy commented:
@brendan-ward I found and fixed the issue with gdal < 3.5.2...

brendan-ward commented on pyogrio/_err.pyx:
```diff
@@ -132,53 +144,79 @@ exception_map = {
     else:
-        # Reformat messages.
-        msg_b = err_msg
+        return
```
Suggest removing the `else` branch entirely

brendan-ward commented on pyogrio/_err.pyx:
```diff
@@ -248,3 +277,170 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+        if err != OGRERR_NONE:
+            stack = self.error_stack.get()
+
+            if stack:
```
```suggestion
            if self.error_stack.get():
```

(and remove line 312, and similar changes below)

brendan-ward commented on pyogrio/_err.pyx:
```diff
@@ -248,3 +277,170 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+        CPLDefaultErrorHandler(err_class, err_no, err_msg)
+        return
+
+    elif err_class == CE_Failure:
```
```suggestion
    if err_class == CE_Failure:
```

(`elif` not needed because above returns)

brendan-ward commented on pyogrio/_err.pyx:
```diff
@@ -248,3 +277,170 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+
+        return
+
+    elif err_class == CE_Warning:
```
```suggestion
    if err_class == CE_Warning:
```

brendan-ward commented on pyogrio/_io.pyx:
```diff
@@ -196,27 +205,27 @@ cdef void* ogr_open(const char* path_c, int mode, char** options) except NULL:
     except NullPointerError:
         raise DataSourceError(
-            "Failed to open dataset (mode={}): {}".format(mode, path_c.decode("utf-8"))
+            f"Failed to open dataset ({mode=}): {path_c.decode('utf-8')}"
```
```suggestion
            f"Failed to open dataset (mode={mode}): {path_c.decode('utf-8')}"
```

theroggy replied:
Why change this, as both are equivalent, but the existing syntax seems more concise ([ref](https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging))?

brendan-ward replied:
Was not familiar with that way of using F-strings!  I thought it was a typo in converting from `.format` to F-string.  Fine to leave as is, then.

brendan-ward reviewed (commented):
Thanks for updating @theroggy ; just a few minor comments.

brendan-ward reviewed (approved):
Thanks @theroggy !
